```json
{
  "CVE-2020-15191": {
    "related": true,
    "summary": "TensorFlow's `dlpack.to_dlpack` function does not properly handle invalid arguments, leading to a null pointer dereference and potential denial of service.",
    "details": {
      "root_cause": "The vulnerability lies in the `dlpack.to_dlpack` function where, upon receiving an invalid argument, the code does not check the status variable, which results in binding references to null pointers and a subsequent crash.",
      "weaknesses": [
        "Improper error handling",
        "Null pointer dereference"
      ],
      "impact": "A remote attacker can cause a denial of service by providing a crafted input that triggers the vulnerability, leading to a crash due to a null pointer dereference.",
      "attack_vectors": [
        "Network",
         "By passing an invalid argument to the `dlpack.to_dlpack` function, a user can trigger the vulnerability."
      ],
      "required_capabilities": "An attacker needs to be able to send a crafted input to the vulnerable `dlpack.to_dlpack` function, however, no special privileges are required."
    },
    "affected_versions": [
      "Tensorflow before versions 2.2.1 and 2.3.1"
    ],
     "fixed_versions": [
      "TensorFlow versions 2.2.1, or 2.3.1"
     ],
    "references": [
      "https://www.suse.com/security/cve/CVE-2020-15191.html",
       "http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15191",
       "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-q8qj-fc9q-cphr",
      "https://github.com/tensorflow/tensorflow/commit/22e07fb204386768e5bcbea563641ea11f96ceb8",
       "https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/thread/TNMEEN772D6LWSHNB64QFB5TB3CZZEF4/"
    ]
  },
   "CVE-2020-15190": {
    "related": true,
    "summary": "TensorFlow's `tf.raw_ops.Switch` operation can cause a segmentation fault due to a null pointer dereference in the eager runtime when processing tensors.",
      "details": {
      "root_cause": "The `tf.raw_ops.Switch` operation outputs two tensors based on a boolean. When the boolean causes one tensor to be empty, the eager runtime traverses all output tensors, including the empty one which is a `nullptr`, resulting in a null pointer dereference and a segmentation fault.",
       "weaknesses": [
        "Null pointer dereference",
         "Improper handling of empty tensors"
      ],
      "impact": "A remote attacker can trigger a segmentation fault, leading to a denial of service.",
       "attack_vectors": [
        "Network",
        "By crafting an input that causes one of the output tensors of the `tf.raw_ops.Switch` operation to be empty, an attacker can trigger the vulnerability."
        ],
      "required_capabilities":"An attacker needs to be able to call the vulnerable `tf.raw_ops.Switch` operation with specific inputs. No special privileges are required."
    },
     "affected_versions": [
       "Tensorflow before versions 1.15.4, 2.0.3, 2.1.2, 2.2.1 and 2.3.1"
      ],
      "fixed_versions": [
        "TensorFlow versions 1.15.4, 2.0.3, 2.1.2, 2.2.1, or 2.3.1"
      ],
    "references": [
      "https://www.suse.com/security/cve/CVE-2020-15190.html",
      "http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15190",
      "https://github.com/tensorflow/tensorflow/releases/tag/v2.3.1",
       "https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/thread/TNMEEN772D6LWSHNB64QFB5TB3CZZEF4/"
    ]
  },
   "CVE-2020-15192": {
    "related": true,
    "summary": "TensorFlow's `dlpack.to_dlpack` function has a memory leak vulnerability when an invalid list of strings is passed due to missing status checks.",
      "details": {
       "root_cause": "When an invalid list of strings is passed to `dlpack.to_dlpack`, the function does not properly check the status of validation failures, which causes a memory leak.",
        "weaknesses": [
          "Memory leak",
          "Improper error handling"
        ],
       "impact": "A remote attacker can cause a denial of service by repeatedly triggering the vulnerability which leads to resource exhaustion due to the memory leak.",
       "attack_vectors": [
        "Network",
         "By providing a crafted input (invalid list of strings) to the `dlpack.to_dlpack` function."
        ],
      "required_capabilities":"An attacker needs to be able to call the vulnerable `dlpack.to_dlpack` operation with specific inputs. No special privileges are required."
    },
       "affected_versions": [
      "Tensorflow before versions 2.2.1 and 2.3.1"
        ],
        "fixed_versions": [
            "TensorFlow versions 2.2.1, or 2.3.1"
          ],
     "references": [
      "https://www.suse.com/security/cve/CVE-2020-15192.html",
      "http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15192",
      "https://github.com/tensorflow/tensorflow/releases/tag/v2.3.1",
       "https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/thread/TNMEEN772D6LWSHNB64QFB5TB3CZZEF4/"
     ]
  },
    "CVE-2020-15193": {
        "related": true,
        "summary": "TensorFlow's `dlpack.to_dlpack` function can be exploited to cause memory corruption by passing a Python object instead of a tensor.",
          "details": {
            "root_cause":"The pybind11 glue code in `dlpack.to_dlpack` assumes that the argument is a tensor. However, users can pass a Python object instead of a tensor, which leads to a `reinterpret_cast` to an `EagerTensor` that fails, using uninitialized memory, resulting in memory corruption.",
            "weaknesses": [
                "Memory corruption",
                "Type confusion",
                "Use of uninitialized memory"
            ],
            "impact": "A remote attacker can cause memory corruption by passing an invalid object leading to unpredictable behavior or a crash.",
           "attack_vectors": [
              "Network",
               "By passing a Python object instead of a tensor to the `dlpack.to_dlpack` function."
            ],
          "required_capabilities": "An attacker needs to be able to pass arbitrary Python objects as input to the vulnerable function, which can be done remotely."
        },
         "affected_versions": [
              "Tensorflow before versions 2.2.1 and 2.3.1"
          ],
          "fixed_versions": [
                "TensorFlow versions 2.2.1, or 2.3.1"
             ],
         "references": [
            "https://www.suse.com/security/cve/CVE-2020-15193.html",
            "http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15193",
            "https://github.com/tensorflow/tensorflow/security/advisories/GHSA-rjjg-hgv6-h69v",
            "https://github.com/tensorflow/tensorflow/commit/22e07fb204386768e5bcbea563641ea11f96ceb8",
             "https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/thread/TNMEEN772D6LWSHNB64QFB5TB3CZZEF4/"
         ]
    },
   "CVE-2020-15194": {
        "related": true,
        "summary": "TensorFlow's `SparseFillEmptyRowsGrad` implementation lacks proper shape validation, leading to a denial of service.",
          "details": {
            "root_cause": "The `SparseFillEmptyRowsGrad` implementation has incomplete validation of the shapes of its arguments. While `reverse_index_map_t` is validated, `grad_values_t` is not, allowing a malicious user to pass a bad `grad_values_t` to trigger an assertion failure in `vec`, leading to denial of service.",
            "weaknesses": [
                "Improper input validation",
                 "Assertion failure"
            ],
            "impact":"A remote attacker can cause a denial of service by providing a crafted input to `SparseFillEmptyRowsGrad` which results in an assertion failure and program termination.",
           "attack_vectors":[
                "Network",
               "By passing an argument with an invalid shape to the `SparseFillEmptyRowsGrad` operation."
            ],
          "required_capabilities": "An attacker needs to be able to provide specific input to the `SparseFillEmptyRowsGrad` operation. No special privileges are required."
        },
         "affected_versions": [
              "Tensorflow before versions 1.15.4, 2.0.3, 2.1.2, 2.2.1 and 2.3.1"
          ],
          "fixed_versions": [
               "TensorFlow versions 1.15.4, 2.0.3, 2.1.2, 2.2.1, or 2.3.1"
            ],
         "references": [
            "https://www.suse.com/security/cve/CVE-2020-15194.html",
            "http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15194",
            "https://github.com/tensorflow/tensorflow/releases/tag/v2.3.1",
            "https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/thread/TNMEEN772D6LWSHNB64QFB5TB3CZZEF4/"
         ]
    }
}
```